File: utils/sampler.js
import { execSync } from 'child_process';
import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';

export const SAMPLE_MODES = ['full', 'first-n', 'highlights'];

export const COST_ESTIMATE = {
  gemini_2_5_flash_per_second: 0.001,
  note: 'Gemini 2.5 Flash ~$0.001/second of video. A 60s video costs ~$0.06.',
};

export function estimateCost(durationSeconds) {
  return {
    duration_seconds: durationSeconds,
    estimated_cost_usd: +(durationSeconds * COST_ESTIMATE.gemini_2_5_flash_per_second).toFixed(4),
    savings_vs_full: null,
  };
}

function getVideoDuration(videoPath) {
  try {
    const output = execSync(
      `ffprobe -v error -show_entries format=duration -of csv="p=0" "${videoPath}"`,
      { encoding: 'utf8' }
    );
    const duration = parseFloat(output.trim());
    if (isNaN(duration) || duration <= 0) {
      throw new Error(`ffprobe returned invalid duration: ${output.trim()}`);
    }
    return duration;
  } catch (err) {
    if (err.message?.includes('ffprobe')) {
      throw new Error('ffprobe not found. Install ffmpeg to use smart sampling.');
    }
    throw err;
  }
}

function clipFirstN(videoPath, durationSeconds) {
  const tempDir = fs.mkdtempSync(path.join(os.tmpdir(), 'vidrev-clip-'));
  const ext = path.extname(videoPath) || '.mp4';
  const clippedPath = path.join(tempDir, `clipped${ext}`);

  console.log(`   ✂️  Clipping first ${durationSeconds}s of video...`);

  execSync(
    `ffmpeg -i "${videoPath}" -t ${durationSeconds} -c copy -avoid_negative_ts make_zero "${clippedPath}" 2>/dev/null`,
    { encoding: 'utf8' }
  );

  if (!fs.existsSync(clippedPath) || fs.statSync(clippedPath).size === 0) {
    execSync(
      `ffmpeg -i "${videoPath}" -t ${durationSeconds} -y "${clippedPath}" 2>/dev/null`,
      { encoding: 'utf8' }
    );
  }

  if (!fs.existsSync(clippedPath)) {
    throw new Error('ffmpeg clipping failed — output file not created');
  }

  const originalSize = fs.statSync(videoPath).size;
  const clippedSize = fs.statSync(clippedPath).size;
  const sizeReduction = ((1 - clippedSize / originalSize) * 100).toFixed(1);

  console.log(`   → Clipped: ${(clippedSize / 1024 / 1024).toFixed(1)} MB (${sizeReduction}% smaller)`);

  return {
    path: clippedPath,
    tempDir,
    mode: 'first-n',
    duration: durationSeconds,
    size_bytes: clippedSize,
  };
}

function extractHighlights(videoPath, targetDuration = 30) {
  const fullDuration = getVideoDuration(videoPath);

  if (fullDuration <= targetDuration) {
    console.log(`   → Video is ${fullDuration.toFixed(1)}s (≤ ${targetDuration}s target) — using full video`);
    return {
      path: videoPath,
      tempDir: null,
      mode: 'full',
      duration: fullDuration,
      size_bytes: fs.statSync(videoPath).size,
    };
  }

  const tempDir = fs.mkdtempSync(path.join(os.tmpdir(), 'vidrev-highlights-'));
  const ext = path.extname(videoPath) || '.mp4';
  const highlightsPath = path.join(tempDir, `highlights${ext}`);

  console.log(`   🎬 Extracting ${targetDuration}s highlight reel from ${fullDuration.toFixed(1)}s video...`);
  console.log(`   → Analyzing motion to find best segments...`);

  const segmentDuration = 5;
  const segmentCount = Math.ceil(targetDuration / segmentDuration);
  const totalSegments = Math.ceil(fullDuration / segmentDuration);

  const motionScores = [];
  for (let i = 0; i < totalSegments; i++) {
    const start = i * segmentDuration;
    const actualDuration = Math.min(segmentDuration, fullDuration - start);

    try {
      const output = execSync(
        `ffmpeg -i "${videoPath}" -ss ${start} -t ${actualDuration} ` +
        `-vf "select='gt(scene,0.1)',metadata=print:file=/dev/stdout" ` +
        `-f null - 2>&1 | grep -c "lavfi" || true`,
        { encoding: 'utf8', shell: '/bin/bash' }
      );
      const score = parseInt(output.trim()) || 1;
      motionScores.push({ start, duration: actualDuration, score });
    } catch {
      motionScores.push({ start, duration: actualDuration, score: 1 });
    }
  }

  const topSegments = motionScores
    .sort((a, b) => b.score - a.score)
    .slice(0, segmentCount)
    .sort((a, b) => a.start - b.start);

  const filterComplex = topSegments
    .map((seg, i) => `[0:v]trim=start=${seg.start}:duration=${seg.duration},setpts=PTS-STARTPTS[v${i}]`)
    .join(';');

  const concatInputs = topSegments.map((_, i) => `[v${i}]`).join('');

  execSync(
    `ffmpeg -i "${videoPath}" -filter_complex "${filterComplex};${concatInputs}concat=n=${topSegments.length}:v=1[outv]" ` +
    `-map "[outv]" -c:v libx264 -preset fast -crf 23 -an -y "${highlightsPath}" 2>/dev/null`,
    { encoding: 'utf8' }
  );

  if (!fs.existsSync(highlightsPath) || fs.statSync(highlightsPath).size === 0) {
    console.log(`   → Motion-based extraction failed, falling back to first-${targetDuration}s...`);
    return clipFirstN(videoPath, targetDuration);
  }

  const originalSize = fs.statSync(videoPath).size;
  const highlightsSize = fs.statSync(highlightsPath).size;
  const sizeReduction = ((1 - highlightsSize / originalSize) * 100).toFixed(1);

  console.log(`   → Highlight reel: ${(highlightsSize / 1024 / 1024).toFixed(1)} MB (${sizeReduction}% smaller)`);
  console.log(`   → Top segments: ${topSegments.map(s => `${s.start.toFixed(0)}s`).join(', ')}`);

  return {
    path: highlightsPath,
    tempDir,
    mode: 'highlights',
    duration: targetDuration,
    size_bytes: highlightsSize,
    segments: topSegments,
  };
}

export function sampleVideo(videoPath, options = {}) {
  const {
    sampleMode = 'full',
    maxDuration = null,
  } = options;

  const fullDuration = getVideoDuration(videoPath);
  const originalSize = fs.statSync(videoPath).size;

  console.log(`\n── Smart Frame Sampling ──`);
  console.log(`   Original: ${fullDuration.toFixed(1)}s, ${(originalSize / 1024 / 1024).toFixed(1)} MB`);
  console.log(`   Mode: ${sampleMode}`);

  if (sampleMode === 'full') {
    console.log(`   → Using full video (no sampling)`);
    return {
      path: videoPath,
      tempDir: null,
      mode: 'full',
      duration: fullDuration,
      size_bytes: originalSize,
      original_duration: fullDuration,
    };
  }

  if (sampleMode === 'first-n') {
    const clipDuration = maxDuration || 30;
    if (clipDuration >= fullDuration) {
      console.log(`   → Requested ${clipDuration}s ≥ video duration ${fullDuration.toFixed(1)}s — using full video`);
      return {
        path: videoPath,
        tempDir: null,
        mode: 'full',
        duration: fullDuration,
        size_bytes: originalSize,
        original_duration: fullDuration,
      };
    }
    return clipFirstN(videoPath, clipDuration);
  }

  if (sampleMode === 'highlights') {
    const targetDuration = maxDuration || 30;
    return extractHighlights(videoPath, targetDuration);
  }

  throw new Error(`Unknown sample mode: ${sampleMode}. Use: ${SAMPLE_MODES.join(', ')}`);
}

export function cleanupSample(sampleResult) {
  if (sampleResult?.tempDir && fs.existsSync(sampleResult.tempDir)) {
    try {
      fs.rmSync(sampleResult.tempDir, { recursive: true, force: true });
      console.log(`   → Cleaned up temporary sample files`);
    } catch (e) {
      console.log(`   → Cleanup warning: ${e.message}`);
    }
  }
}

File: utils/sampler.py
from __future__ import annotations

import math
import os
import shutil
import subprocess
import tempfile
from typing import Any

SAMPLE_MODES = ["full", "first-n", "highlights"]


def _get_video_duration(video_path: str) -> float:
    try:
        result = subprocess.run(
            ["ffprobe", "-v", "error", "-show_entries", "format=duration", "-of", "csv=p=0", video_path],
            capture_output=True,
            text=True,
            check=True,
        )
        duration = float(result.stdout.strip())
        if math.isnan(duration) or duration <= 0:
            raise ValueError(f"ffprobe returned invalid duration: {result.stdout.strip()}")
        return duration
    except FileNotFoundError as err:
        raise RuntimeError("ffprobe not found. Install ffmpeg to use smart sampling.") from err
    except subprocess.CalledProcessError as e:
        raise RuntimeError(f"ffprobe failed: {e.stderr}") from e


def _clip_first_n(video_path: str, duration_seconds: float) -> dict[str, Any]:
    temp_dir = tempfile.mkdtemp(prefix="vidrev-clip-")
    ext = os.path.splitext(video_path)[1] or ".mp4"
    clipped_path = os.path.join(temp_dir, f"clipped{ext}")

    print(f"   ✂️  Clipping first {duration_seconds}s of video...", flush=True)

    try:
        subprocess.run(
            [
                "ffmpeg",
                "-i",
                video_path,
                "-t",
                str(duration_seconds),
                "-c",
                "copy",
                "-avoid_negative_ts",
                "make_zero",
                clipped_path,
            ],
            capture_output=True,
            text=True,
            check=True,
        )
    except subprocess.CalledProcessError:
        subprocess.run(
            ["ffmpeg", "-i", video_path, "-t", str(duration_seconds), "-y", clipped_path],
            capture_output=True,
            text=True,
            check=True,
        )

    if not os.path.exists(clipped_path) or os.path.getsize(clipped_path) == 0:
        raise RuntimeError("ffmpeg clipping failed — output file not created")

    original_size = os.path.getsize(video_path)
    clipped_size = os.path.getsize(clipped_path)
    size_reduction = (1 - clipped_size / original_size) * 100

    print(f"   → Clipped: {clipped_size / 1024 / 1024:.1f} MB ({size_reduction:.1f}% smaller)", flush=True)

    return {
        "path": clipped_path,
        "temp_dir": temp_dir,
        "mode": "first-n",
        "duration": duration_seconds,
        "size_bytes": clipped_size,
    }


def _extract_highlights(video_path: str, target_duration: float = 30) -> dict[str, Any]:
    full_duration = _get_video_duration(video_path)

    if full_duration <= target_duration:
        print(f"   → Video is {full_duration:.1f}s (≤ {target_duration}s target) — using full video", flush=True)
        return {
            "path": video_path,
            "temp_dir": None,
            "mode": "full",
            "duration": full_duration,
            "size_bytes": os.path.getsize(video_path),
        }

    temp_dir = tempfile.mkdtemp(prefix="vidrev-highlights-")
    ext = os.path.splitext(video_path)[1] or ".mp4"
    highlights_path = os.path.join(temp_dir, f"highlights{ext}")

    print(f"   🎬 Extracting {target_duration}s highlight reel from {full_duration:.1f}s video...", flush=True)
    print("   → Analyzing motion to find best segments...", flush=True)

    segment_duration = 5
    segment_count = math.ceil(target_duration / segment_duration)
    total_segments = math.ceil(full_duration / segment_duration)

    motion_scores = []
    for i in range(total_segments):
        start = i * segment_duration
        actual_duration = min(segment_duration, full_duration - start)

        try:
            result = subprocess.run(
                [
                    "ffmpeg",
                    "-i",
                    video_path,
                    "-ss",
                    str(start),
                    "-t",
                    str(actual_duration),
                    "-vf",
                    "select=gt(scene\\,0.1),metadata=print:file=/dev/stdout",
                    "-f",
                    "null",
                    "-",
                ],
                capture_output=True,
                text=True,
            )
            score = result.stderr.count("lavfi")
            if score == 0:
                score = 1
            motion_scores.append({"start": start, "duration": actual_duration, "score": score})
        except Exception:
            motion_scores.append({"start": start, "duration": actual_duration, "score": 1})

    top_segments = sorted(motion_scores, key=lambda x: x["score"], reverse=True)[:segment_count]
    top_segments = sorted(top_segments, key=lambda x: x["start"])

    filter_parts = []
    concat_inputs = []
    for i, seg in enumerate(top_segments):
        filter_parts.append(f"[0:v]trim=start={seg['start']}:duration={seg['duration']},setpts=PTS-STARTPTS[v{i}]")
        concat_inputs.append(f"[v{i}]")

    filter_complex = ";".join(filter_parts) + ";" + "".join(concat_inputs) + f"concat=n={len(top_segments)}:v=1[outv]"

    try:
        subprocess.run(
            [
                "ffmpeg",
                "-i",
                video_path,
                "-filter_complex",
                filter_complex,
                "-map",
                "[outv]",
                "-c:v",
                "libx264",
                "-preset",
                "fast",
                "-crf",
                "23",
                "-an",
                "-y",
                highlights_path,
            ],
            capture_output=True,
            text=True,
            check=True,
        )
    except subprocess.CalledProcessError:
        print("   → Motion-based extraction failed, falling back to first segment...", flush=True)
        return _clip_first_n(video_path, target_duration)

    if not os.path.exists(highlights_path) or os.path.getsize(highlights_path) == 0:
        print("   → Motion-based extraction failed, falling back to first segment...", flush=True)
        return _clip_first_n(video_path, target_duration)

    original_size = os.path.getsize(video_path)
    highlights_size = os.path.getsize(highlights_path)
    size_reduction = (1 - highlights_size / original_size) * 100

    print(f"   → Highlight reel: {highlights_size / 1024 / 1024:.1f} MB ({size_reduction:.1f}% smaller)", flush=True)
    seg_strs = [f"{s['start']:.0f}s" for s in top_segments]
    print(f"   → Top segments: {', '.join(seg_strs)}", flush=True)

    return {
        "path": highlights_path,
        "temp_dir": temp_dir,
        "mode": "highlights",
        "duration": target_duration,
        "size_bytes": highlights_size,
        "segments": top_segments,
    }


def sample_video(video_path: str, options: dict[str, Any] | None = None) -> dict[str, Any]:
    if options is None:
        options = {}

    sample_mode = options.get("sample_mode", "full")
    max_duration = options.get("max_duration")

    full_duration = _get_video_duration(video_path)
    original_size = os.path.getsize(video_path)

    print("\n── Smart Frame Sampling ──", flush=True)
    print(f"   Original: {full_duration:.1f}s, {original_size / 1024 / 1024:.1f} MB", flush=True)
    print(f"   Mode: {sample_mode}", flush=True)

    if sample_mode == "full":
        print("   → Using full video (no sampling)", flush=True)
        return {
            "path": video_path,
            "temp_dir": None,
            "mode": "full",
            "duration": full_duration,
            "size_bytes": original_size,
            "original_duration": full_duration,
        }

    if sample_mode == "first-n":
        clip_duration = max_duration if max_duration else 30
        if clip_duration >= full_duration:
            print(
                f"   → Requested {clip_duration}s ≥ video duration {full_duration:.1f}s — using full video", flush=True
            )
            return {
                "path": video_path,
                "temp_dir": None,
                "mode": "full",
                "duration": full_duration,
                "size_bytes": original_size,
                "original_duration": full_duration,
            }
        return _clip_first_n(video_path, clip_duration)

    if sample_mode == "highlights":
        target_duration = max_duration if max_duration else 30
        return _extract_highlights(video_path, target_duration)

    raise ValueError(f"Unknown sample mode: {sample_mode}. Use: {', '.join(SAMPLE_MODES)}")


def cleanup_sample(sample_result: dict[str, Any] | None) -> None:
    if sample_result and sample_result.get("temp_dir") and os.path.exists(sample_result["temp_dir"]):
        try:
            shutil.rmtree(sample_result["temp_dir"], ignore_errors=True)
            print("   → Cleaned up temporary sample files", flush=True)
        except Exception as e:
            print(f"   → Cleanup warning: {e}", flush=True)
